src/backend/distributed/commands/multi_copy.c

```diff
@@ -3509,6 +3555,51 @@ CopyGetPlacementConnection(ShardPlacement *placement, bool stopOnFailure)
 }
 
 
+static MultiConnection *
+GetConnectionIfExecutorPoolSizeReached(HTAB *connectionStateHash, char *nodeName, int
+									   nodePort)
+{
+	List *connectionStateList =
+		ConnectionStateListToNode(connectionStateHash, nodeName, nodePort);
+
+	if (list_length(connectionStateList) < MaxAdaptiveExecutorPoolSize)
+	{
+		return NULL;
+	}
+
+	MultiConnection *connection = NULL;
+	int minPlacementCount = INT32_MAX;
+	ListCell *connectionStateCell = NULL;
+
+	foreach(connectionStateCell, connectionStateList)
+	{
+		CopyConnectionState *connectionState = lfirst(connectionStateCell);
+		int currentConnectionPlacementCount = 0;
+
+		if (connectionState->activePlacementState != NULL)
+		{
+			currentConnectionPlacementCount++;
```

Comment:
we can also maintain this `CopyConnectionState->placementCount` to avoid the loop over `bufferedPlacementList` here, and make the code a bit nicer here.